core/src/main/java/hudson/security/SecurityRealm.java

```diff
@@ -484,6 +486,57 @@ public Filter createFilter(FilterConfig filterConfig) {
      */
     public static final SecurityRealm NO_AUTHENTICATION = new None();
 
+    /**
+     * Perform a calculation where we should go back after sucessfull login
+     *
+     * @return Encoded URI where we should go back after sucessfull login
+     *         or "/" if no way back or an issue occurred
+     *
+     * @since TODO
+     */
+    public static String getFrom() {
+        String from = null, returnValue = null;
+        final StaplerRequest request = Stapler.getCurrentRequest();
+
+        // Try to obtain a return point either from the Session
+        // or from the QueryParameter in this order
+        if (request != null
+                && request.getSession(false) != null) {
+            from = (String) request.getSession().getAttribute("from");
+        } else {
+            if (request != null) {
+                from = request.getParameter("from");
+            }
+        }
+
+        // If entry point was not found, try to deduce it from the request URI
+        // except pages related to login process
+        if (from == null
+                && request != null
+                && request.getRequestURI() != null
+                && request.getRequestURI().compareTo("/loginError") != 0
+                && request.getRequestURI().compareTo("/login") != 0) {
+
+                from = request.getRequestURI();
+        }
+
+        // If deduced entry point isn't deduced yet or the content is a blank value
+        // use the root web point "/" as a fallback
+        from.trim();
```

Comment:
My bad, too long to look inside it...:) trim() should be after the condition, fixing...